crates/ursa-network/src/behaviour.rs

```diff
@@ -660,71 +696,76 @@ impl<P: StoreParams> Behaviour<P> {
-        self.handle_ping(event)
-    }
-}
+impl<P: StoreParams> NetworkBehaviour for Behaviour<P> {
+    type ConnectionHandler = <InternalBehaviour<P> as NetworkBehaviour>::ConnectionHandler;
+    type OutEvent = BehaviourEvent;
 
-impl<P: StoreParams> NetworkBehaviourEventProcess<IdentifyEvent> for Behaviour<P> {
-    fn inject_event(&mut self, event: IdentifyEvent) {
-        self.handle_identify(event)
+    fn new_handler(&mut self) -> Self::ConnectionHandler {
+        self.inner.new_handler()
     }
-}
 
-impl<P: StoreParams> NetworkBehaviourEventProcess<GossipsubEvent> for Behaviour<P> {
-    fn inject_event(&mut self, event: GossipsubEvent) {
-        self.handle_gossipsub(event)
+    fn addresses_of_peer(&mut self, peer: &PeerId) -> Vec<Multiaddr> {
+        self.inner.addresses_of_peer(peer)
     }
-}
 
-impl<P: StoreParams> NetworkBehaviourEventProcess<BitswapEvent> for Behaviour<P> {
-    fn inject_event(&mut self, event: BitswapEvent) {
-        self.handle_bitswap(event)
+    fn on_swarm_event(&mut self, event: FromSwarm<Self::ConnectionHandler>) {
+        self.inner.on_swarm_event(event);
     }
-}
-
-impl<P: StoreParams> NetworkBehaviourEventProcess<DiscoveryEvent> for Behaviour<P> {
-    fn inject_event(&mut self, event: DiscoveryEvent) {
-        self.handle_discovery(event)
-    }
-}
-
-impl<P: StoreParams> NetworkBehaviourEventProcess<AutonatEvent> for Behaviour<P> {
-    fn inject_event(&mut self, event: AutonatEvent) {
-        self.handle_autonat(event)
-    }
-}
-
-impl<P: StoreParams> NetworkBehaviourEventProcess<RelayServerEvent> for Behaviour<P> {
-    fn inject_event(&mut self, event: RelayServerEvent) {
-        self.handle_relay_server(event)
-    }
-}
-
-impl<P: StoreParams> NetworkBehaviourEventProcess<RelayClientEvent> for Behaviour<P> {
-    fn inject_event(&mut self, event: RelayClientEvent) {
-        self.handle_relay_client(event)
-    }
-}
 
-impl<P: StoreParams> NetworkBehaviourEventProcess<DcutrEvent> for Behaviour<P> {
-    fn inject_event(&mut self, event: DcutrEvent) {
-        self.handle_dcutr(event)
+    fn on_connection_handler_event(
+        &mut self,
+        peer_id: PeerId,
+        connection_id: ConnectionId,
+        event: <<Self::ConnectionHandler as IntoConnectionHandler>::Handler as ConnectionHandler>::OutEvent,
+    ) {
+        self.inner
+            .on_connection_handler_event(peer_id, connection_id, event)
     }
-}
 
-impl<P: StoreParams>
-    NetworkBehaviourEventProcess<RequestResponseEvent<UrsaExchangeRequest, UrsaExchangeResponse>>
-    for Behaviour<P>
-{
-    fn inject_event(
+    fn poll(
         &mut self,
-        event: RequestResponseEvent<UrsaExchangeRequest, UrsaExchangeResponse>,
-    ) {
-        self.handle_request_response(event)
+        cx: &mut Context<'_>,
+        params: &mut impl PollParameters,
+    ) -> Poll<NetworkBehaviourAction<Self::OutEvent, Self::ConnectionHandler>> {
+        loop {
+            if let Some(event) = self.events.pop_front() {
```

Comment:
So the service uses `publish_ad` to queue events so here we dequeue them and return. Let me know if you think this is fine. @theBeardA Thanks for catching that.

Reply:
One concern with this is that events pushed by `publish_ad` could pass ahead of the queue and not let us poll the other inner behaviours. Hmm...

Reply:
I don't see a function in `libp2p` API that could inject an event so it can be polled returned along with the other events. Another option is to make a "dummy": NetworkBehaviour struct that can be polled. Internally it will have the `events` queue. Thoughts?

Reply:
I was thinking something like this.

```rust
struct Dummy {
  events: VecDeque
}

impl NetworkBehaviour for Dummy {
...
    fn poll() {
      if let Some(event) = self.events.pop_front() {
        return Poll::Ready(event)
      } 
      Poll::pending
    }
}
```

Reply:
In this way, it will be polled "fairly"

Reply:
Awesome. this makes sense. thanks

Reply:
This might not be needed, after discussing with @b0xtch, these functions will get refactored and handling will be moved to the service.